README.md
# sqtom - Squeezed state tomography in Python

![GitHub Workflow Status (master)](https://img.shields.io/github/workflow/status/XanaduAI/sqtom/Tests/master?style=flat-square)
![Codecov coverage](https://img.shields.io/codecov/c/github/xanaduai/sqtom/master.svg?style=flat-square)
![CodeFactor Grade](https://img.shields.io/codefactor/grade/github/XanaduAI/sqtom/master?style=flat-square)
![PyPI](https://img.shields.io/pypi/v/sqtom.svg?style=flat-square)
![PyPI - Python Version](https://img.shields.io/pypi/pyversions/sqtom.svg?style=flat-square)

This repository implements the mode tomography ideas presented in

*"Full statistical mode reconstruction of a light field via a photon-number-resolved measurement"*
by Burenkov. et al. [Phys. Rev. A 95, 053806 (2017)](https://journals.aps.org/pra/abstract/10.1103/PhysRevA.95.053806)
and in Burenkov et al. in [J. Res. Natl. Inst. Stan. 122, 30 (2017)](https://doi.org/10.6028/jres.122.030).


for twin beam light and extends it to degenerate squeezed light. By leveraging `lmfit` we can also
give a number of uncertainty estimates, and moreover provide routines for thresholding photon-number
measurements and useful heuristics for initial guesses for the solutions of the problem.

## Contents

The main physical ideal used by Burenkov et al. is to model the joint photon distribution of the
variables associated to the photon numbers in signal and idler beams as resulting from one or
several lossy two-mode squeezed distributions hitting the detectors. To model dark counts they also
allow for modes prepared in states with Poisson statistics to hit the detectors.

To obtain the joint probability distribution of the photon numbers in the signal and idlers one
needs to *convolve* the probability distributions of the modes entering in the problem.

## Requirements

* [SciPy](https://www.scipy.org/) to calculate probability distributions of Poisson, Geometric or
  Negative Binomial random variables.

* [NumPy](https://numpy.org/) to perform 2D convolutions and matrix manipulations.

* [The Walrus](https://the-walrus.readthedocs.io/en/latest/) to calculate loss matrices and squeezed
  states probability distributions.

With the tools described so far we can solve the *forward* problem, i.e., given a set of physical
parameters what is the probability distribution.

* If we augment our tools with [lmfit](https://lmfit.github.io/lmfit-py/) we can solve the *inverse*
  problem: to find the best set of parameters that explain a given observed frequency distribution
  of photon numbers.

Finally, we use [pytest](https://docs.pytest.org/en/latest/) for testing.

All of these prerequisites can be installed via `pip`:

```bash
pip install sqtom
```

## Author

Nicolas Quesada

## License

This source code is free and open source, released under the Apache License, Version 2.0.


.github/CHANGELOG.md
# Version 0.4.0-dev

### New features

* New function for the the generation of reasonable parameter guesses for degenerate squeezing. [#17](https://github.com/XanaduAI/sqtom/pull/17)

* New convenience function for calculating 1D histograms. [#17](https://github.com/XanaduAI/sqtom/pull/17)

* Adds the third order correlation function `g3` into `marginal_calcs_1d` and `marginal_calcs_2d`. [#18](https://github.com/XanaduAI/sqtom/pull/18)

### Improvements

### Bug fixes

### Breaking changes

### Contributors

This release contains contributions from (in alphabetical order):

David Phillips



sqtom/tests/test_fitting_1d.py
# Copyright 2019-2020 Xanadu Quantum Technologies Inc.

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#     http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Basic tests for the functions in fitting_1d"""

import pytest
import numpy as np
from sqtom.forward_solver import degenerate_pmf
from sqtom.fitting_1d import (
    two_schmidt_mode_guess,
    marginal_calcs_1d,
    gen_hist_1d,
    threshold_1d,
    fit_1d,
)


@pytest.mark.parametrize("sq_0", [0.1, 0.5, 1.0])
def test_gen_hist_1d(sq_0):
    """Check that a histogram is constructed correctly for a degenerate squeezing source"""
    nsamples = 1_000_000
    q = 1.0 - np.tanh(np.arcsinh(np.sqrt(sq_0))) ** 2
    r = 0.5
    samples = 2 * np.random.negative_binomial(r, q, size=nsamples)
    nmax = max(samples)
    expected_pmf = degenerate_pmf({"sq_0": sq_0, "n_modes": 1}, cutoff=nmax)
    pmf = gen_hist_1d(samples)
    atol = 5 / np.sqrt(nsamples)
    assert np.allclose(pmf, expected_pmf, atol=atol)


@pytest.mark.parametrize("eta", [0.1, 0.5, 1.0])
@pytest.mark.parametrize("sq_0", [0.1, 1.0, 2.0])
@pytest.mark.parametrize("sq_1", [0.0, 0.1, 1.0])
def test_two_schmidt_mode_guess_exact(eta, sq_0, sq_1):
    """Test that one can invert correctly when there are two Schmidt modes
    and no dark counts
    """
    pmf = degenerate_pmf({"sq_0": sq_0, "sq_1": sq_1, "n_modes": 2, "eta": eta})
    guess = two_schmidt_mode_guess(pmf)
    assert np.allclose(eta, guess["eta"], atol=1.0e-2)
    sq_ns = [sq_0, sq_1]  # We need to sort sq_n0 and sq_n1 so that sq_n0 >= sq_n1
    sq_0 = np.max(sq_ns)
    sq_1 = np.min(sq_ns)
    assert np.allclose(sq_0, guess["sq_0"], atol=0.1)
    assert np.allclose(sq_1, guess["sq_1"], atol=0.1)


@pytest.mark.parametrize("pos", [0, 1])
def test_two_schmidt_modes_guess_warning(pos):
    """Tests None is returned if the mean photon number or g2 are zero"""
    probs = np.zeros([10])
    probs[pos] = 1.0
    guess = two_schmidt_mode_guess(probs)
    assert guess is None


@pytest.mark.parametrize("do_not_vary", ["eta", "noise", None])
@pytest.mark.parametrize("n_modes", [1, 2, 3])
@pytest.mark.parametrize("threshold", [False, 10])
def test_exact_model_1d(n_modes, threshold, do_not_vary):
    """Test that the fitting is correct when the guess is exactly the correct answer"""
    sq_n = 0.7 * (0.5 ** np.arange(n_modes))
    noise = 0.1
    eta = 0.7
    params = {"sq_" + str(i): sq_n[i] for i in range(n_modes)}
    params["n_modes"] = n_modes
    params["eta"] = eta
    params["noise"] = noise
    if threshold:
        probs = threshold_1d(degenerate_pmf(params), threshold)
    else:
        probs = degenerate_pmf(params)
    fit = fit_1d(probs, params, threshold=threshold, do_not_vary=do_not_vary)
    assert np.allclose(fit.chisqr, 0.0)


def test_marginal_calcs_1d():
    """Tests that marginal_calcs_1d returns the correct values as an array"""
    nmean = 1.0
    ps = degenerate_pmf({"n_modes": 1, "sq_0": nmean}, cutoff=100)
    assert np.allclose(
        marginal_calcs_1d(ps, as_dict=False), np.array([nmean, 3 + 1 / nmean, 15 + 9 / nmean])
    )


sqtom/tests/test_fitting_2d.py
# Copyright 2019-2020 Xanadu Quantum Technologies Inc.

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#     http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Basic tests for the functions in fitting_2d"""

import pytest
import numpy as np
from sqtom.forward_solver import twinbeam_pmf
from sqtom.fitting_2d import (
    fit_2d,
    gen_hist_2d,
    two_schmidt_mode_guess,
    marginal_calcs_2d,
    threshold_2d,
)


@pytest.mark.parametrize("sq_0", [0.1, 1.0, 2.0])
def test_gen_hist_2d_twin(sq_0):
    """Check that a histogram is constructed correctly for a lossless pure twin-beam source"""
    nsamples = 1000000
    p = 1 / (1.0 + sq_0)
    samples = np.random.geometric(p, nsamples) - 1
    mat = gen_hist_2d(samples, samples)
    n, m = mat.shape
    assert n == m
    expected = twinbeam_pmf({"sq_0": sq_0, "n_modes": 1}, cutoff=n)
    assert np.allclose(mat, expected, atol=0.01)


@pytest.mark.parametrize("ns", [0.1, 1.0, 2.0])
@pytest.mark.parametrize("ni", [0.1, 1.0, 2.0])
def test_gen_hist_2d_poisson(ns, ni):
    """Test the histograms are correctly generated for pure noise"""
    nsamples = 1000000
    mat = gen_hist_2d(np.random.poisson(ns, nsamples), np.random.poisson(ni, nsamples))
    n, m = mat.shape
    nmax = np.max([n, m])
    expected = twinbeam_pmf({"noise_s": ns, "noise_i": ni})[:n, :m]
    np.allclose(expected, mat, atol=0.01)


@pytest.mark.parametrize("eta_s", [0.1, 0.5, 1.0])
@pytest.mark.parametrize("eta_i", [0.1, 0.5, 1.0])
@pytest.mark.parametrize("sq_0", [0.0, 0.1, 1.0, 2.0])
@pytest.mark.parametrize("sq_1", [0.1, 1.0, 2.0])
def test_two_schmidt_mode_guess_exact(eta_s, eta_i, sq_0, sq_1):
    """Test that one can invert correctly when there are two Schmidt modes
    and no dark counts.
    """
    pmf = twinbeam_pmf({"sq_0": sq_0, "sq_1": sq_1, "n_modes": 2, "eta_s": eta_s, "eta_i": eta_i})
    guess = two_schmidt_mode_guess(pmf)
    assert np.allclose(eta_s, guess["eta_s"], atol=1.0e-2)
    assert np.allclose(eta_i, guess["eta_i"], atol=1.0e-2)
    sq_ns = [sq_0, sq_1]  # We need to sort sq_n1 and sq_n2 so that sq_n1 >= sq_n2
    sq_0 = np.max(sq_ns)
    sq_1 = np.min(sq_ns)
    assert np.allclose(sq_0, guess["sq_0"], atol=1.0e-2)
    assert np.allclose(sq_1, guess["sq_1"], atol=1.0e-2)


@pytest.mark.parametrize("do_not_vary", ["eta_s", "noise_s", "eta_i", "noise_i", None])
@pytest.mark.parametrize("n_modes", [1, 2, 3])
@pytest.mark.parametrize("threshold", [False, 5])
def test_exact_model_2d(n_modes, do_not_vary, threshold):
    """Test that the fitting is correct when the guess is exactly the correct answer"""
    sq_n = 0.7 * (0.5 ** np.arange(n_modes))
    noise_s = 0.1
    noise_i = 0.15
    eta_s = 0.7
    eta_i = 0.6
    params = {"sq_" + str(i): sq_n[i] for i in range(n_modes)}
    params["n_modes"] = n_modes
    params["eta_s"] = eta_s
    params["eta_i"] = eta_i
    params["noise_s"] = noise_s
    params["noise_i"] = noise_i
    if threshold:
        probs = threshold_2d(twinbeam_pmf(params), threshold, threshold)
    else:
        probs = twinbeam_pmf(params)
    fit = fit_2d(probs, params, do_not_vary=do_not_vary, threshold=threshold)
    assert np.allclose(fit.chisqr, 0.0)


def test_marginal_calcs_2d():
    """Tests that marginal_calcs_2d returns the correct values as an array"""
    nmean = 1.0
    ps = twinbeam_pmf({"n_modes": 1.0, "sq_0": nmean})
    res = marginal_calcs_2d(ps, as_dict=False)
    expected = np.array([nmean, nmean, 2 + 1 / nmean, 2, 2, 0, 6, 6])
    assert np.allclose(res, expected)


sqtom/tests/test_photon_tom.py
from sqtom import twinbeam_pmf, loss_mat, degenerate_pmf, marginal_calcs_2d, two_schmidt_mode_guess, marginal_calcs_1d, gen_hist_2d
import pytest
import numpy as np
from scipy.stats import geom


@pytest.mark.parametrize("eta_s", [0.1, 0.5, 1.0])
@pytest.mark.parametrize("eta_i", [0.1, 0.5, 1.0])
@pytest.mark.parametrize("sq_n", [0.1, 1.0, 2.0])
def test_marginal_calcs_one_schmidt_mode(eta_s, eta_i, sq_n):
    """ Check that the correct g2s and g11 and mean photon numbers
    are calculated when a single Schmidt mode is involved.
    """
    n = 30
    pmf = twinbeam_pmf(n, eta_s=eta_s, eta_i=eta_i, twin_bose=[sq_n])
    marginals = marginal_calcs_2d(pmf)
    assert np.allclose(marginals["g2s"], 2.0, rtol=0.001)
    assert np.allclose(marginals["g2i"], 2.0, rtol=0.001)
    assert np.allclose(marginals["g11"], 2.0 + 1.0 / sq_n, rtol=0.001)
    assert np.allclose(marginals["ns"], eta_s * sq_n, rtol=0.001)
    assert np.allclose(marginals["ni"], eta_i * sq_n, rtol=0.001)


@pytest.mark.parametrize("sq_n", [0.1, 1.0, 2.0])
def test_gen_hist_2d_twin(sq_n):
    """Check that a histogram is constructed correctly for a lossless pure twin-beam source"""
    nsamples = 1000000
    p = 1 / (1.0 + sq_n)
    samples = np.random.geometric(p, nsamples) - 1
    samples2d = np.array([samples, samples])
    mat = gen_hist_2d(samples2d[0], samples2d[1])
    n, m = mat.shape
    assert n == m
    expected = twinbeam_pmf(n, twin_bose=[sq_n])
    assert np.allclose(mat, expected, atol=0.01)


@pytest.mark.parametrize("ns", [0.1, 1.0, 2.0])
@pytest.mark.parametrize("ni", [0.1, 1.0, 2.0])
def test_gen_hist_2d_poisson(ns, ni):
    nsamples = 1000000
    samples2d = np.array(
        [np.random.poisson(ns, nsamples), np.random.poisson(ni, nsamples)]
    )
    mat = gen_hist_2d(samples2d[0], samples2d[1])
    n, m = mat.shape
    nmax = np.max([n, m])
    expected = twinbeam_pmf(nmax, poisson_param_ns=ns, poisson_param_ni=ni)[:n, :m]
    np.allclose(expected, mat, atol=0.01)


@pytest.mark.parametrize("eta", [0.1, 0.5, 1.0])
def test_loss_is_stochastic_matrix(eta):
    """Test the loss matrix is an stochastic matrix, implying that the sum
    of the entries a long the rows is 1"""
    n = 50
    M = loss_mat(eta, n)
    assert np.allclose(np.sum(M, axis=1), np.ones([n]))


@pytest.mark.parametrize("eta", [0.1, 0.5, 1.0])
def test_loss_is_nonnegative_matrix(eta):
    """Test the loss matrix is a nonnegative matrix"""
    n = 50
    M = loss_mat(eta, n)
    assert np.alltrue(M >= 0.0)


@pytest.mark.parametrize("eta", [-1.0, 2.0])
def test_value_error(eta):
    """Tests the correct error is raised"""
    n = 50
    with pytest.raises(
        ValueError,
        match="The transmission parameter eta should be a number between 0 and 1.",
    ):
        loss_mat(eta, n)


@pytest.mark.parametrize("eta_s", [0.1, 0.5, 1.0])
@pytest.mark.parametrize("eta_i", [0.1, 0.5, 1.0])
@pytest.mark.parametrize("sq_n1", [0.0, 0.1, 1.0, 2.0])
@pytest.mark.parametrize("sq_n2", [0.1, 1.0, 2.0])
def test_two_schmidt_mode_guess_exact(eta_s, eta_i, sq_n1, sq_n2):
    """Test that one can invert correctly when there are two Schmidt modes
    and no dark counts.
    """
    nmax = 40
    pmf = twinbeam_pmf(nmax, eta_s=eta_s, eta_i=eta_i, twin_bose=[sq_n1, sq_n2])
    guess = two_schmidt_mode_guess(pmf)
    assert np.allclose(eta_s, guess["etas"], atol=1.0e-2)
    assert np.allclose(eta_i, guess["etai"], atol=1.0e-2)
    sq_ns = [sq_n1, sq_n2]  # We need to sort sq_n1 and sq_n2 so that sq_n1 >= sq_n2
    sq_n1 = np.max(sq_ns)
    sq_n2 = np.min(sq_ns)
    assert np.allclose(sq_n1, guess["twin_n1"], atol=1.0e-2)
    assert np.allclose(sq_n2, guess["twin_n2"], atol=1.0e-2)


@pytest.mark.parametrize("eta_s", [0.1, 0.5, 1.0])
@pytest.mark.parametrize("eta_i", [0.1, 0.5, 1.0])
@pytest.mark.parametrize("sq_n1", [0.0, 0.1, 1.0, 2.0])
@pytest.mark.parametrize("sq_n2", [0.1, 1.0, 2.0])
@pytest.mark.parametrize("dc_s", [0.2, 0.5])
@pytest.mark.parametrize("dc_i", [0.2, 0.5])
def test_dark_counts_g2_twin(sq_n1, sq_n2, dc_s, dc_i, eta_s, eta_i):
    """Test that dark counts are correctly included by updating the dark counts
    """
    nmax = 40
    pmf = twinbeam_pmf(
        nmax,
        eta_s=eta_s,
        eta_i=eta_i,
        poisson_param_ns=dc_s,
        poisson_param_ni=dc_i,
        twin_bose=[sq_n1, sq_n2],
    )
    K = (sq_n1 + sq_n2) ** 2 / (sq_n1 ** 2 + sq_n2 ** 2)
    M = sq_n1 + sq_n2
    g2noiseless = 1.0 + 1.0 / K
    eps_s = dc_s / M
    eps_i = dc_i / M
    g2s = (g2noiseless + 2 * eps_s + eps_s ** 2) / (1.0 + 2 * eps_s + eps_s ** 2)
    g2i = (g2noiseless + 2 * eps_i + eps_i ** 2) / (1.0 + 2 * eps_i + eps_i ** 2)
    marginals = marginal_calcs_2d(pmf)
    np.allclose(g2s, marginals["g2s"])
    np.allclose(g2i, marginals["g2i"])
    np.allclose(eta_s * (dc_s + sq_n1 + sq_n2), marginals["ns"])
    np.allclose(eta_i * (dc_i + sq_n1 + sq_n2), marginals["ns"])


@pytest.mark.parametrize("eta", [0.1, 0.5, 1.0])
@pytest.mark.parametrize("sq_n1", [0.0, 0.1, 1.0, 2.0])
@pytest.mark.parametrize("sq_n2", [0.1, 1.0, 2.0])
@pytest.mark.parametrize("dc", [0.2, 0.5])
def test_g2_degenerate(eta, sq_n1, sq_n2, dc):
    """Test that the g2 of a single mode degenerate squeezer is 3+1/n regardless of the loss, where n is the mean photon number"""
    nmax = 50
    ps = degenerate_pmf(nmax, sq_n=[sq_n1, sq_n2], eta=eta, n_dark=eta * dc)
    vals = marginal_calcs_1d(ps)
    K = (sq_n1 + sq_n2) ** 2 / (sq_n1 ** 2 + sq_n2 ** 2)
    M = sq_n1 + sq_n2
    g2noiseless = 1.0 + 1.0 / M + 2.0 / K
    eps = dc / M
    g2 = (g2noiseless + 2 * eps + eps ** 2) / (1 + 2 * eps + eps ** 2)
    assert np.allclose(vals["g2"], g2, atol=0.05)
    assert np.allclose(vals["n"], eta * (M + dc), atol=0.05)


@pytest.mark.parametrize("sq_n", [0.1, 1.0, 2.0])
def test_pmf_two_schmidt_degenerate(sq_n):
    """Test that the photon number dist of a degenerate squeezer with two Schmidt mode is a geometric distribution in the pair number."""
    nmax = 50
    ps = geom.pmf(np.arange(1, nmax + 1), 1.0 / (1.0 + sq_n))
    expected = np.zeros([2 * nmax])
    expected[0::2] = ps
    ps = degenerate_pmf(nmax, sq_n=[sq_n, sq_n])
    assert np.allclose(ps, expected[0:nmax])


sqtom/__init__.py
# Copyright 2019-2020 Xanadu Quantum Technologies Inc.

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#     http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
r""" Code for mode tomography"""
from ._version import __version__


def version():  # pragma: no cover
    r"""
    Get version number of sqtom
    Returns:
      str: The package version number
    """
    return __version__


sqtom/_fitting_1d.py
import numpy as np
from lmfit import Minimizer, Parameters
from ._photon_tom import degenerate_pmf


def marginal_calcs_1d(pd_data, as_dict=True):
    """ Given a two dimensional array of probabilities it calculates the first
    two moments of the marginal distributions and also their g2's and g11
    It returns these values as a dictionary"""

    (intn,) = pd_data.shape
    n = np.arange(intn)
    nmean = pd_data @ n
    nmean2 = pd_data @ n ** 2
    g2 = (nmean2 - nmean) / nmean ** 2
    if as_dict is True:
        return {"n": nmean, "g2": g2}
    return np.array([nmean, g2])


def model_1d(params, pd_data, n_max=50):
    """Constructs a joint probability distribution (jpd) given squeezer parameters
    like noise, squeezing values and noise and the returns the difference between
    the constructed jpd and the the jpd of the data that is to be fit.
    Args:
        params (dict): dictionary of all the Parameter objects required to specify a fit model
        pd_data (array): rectangular array with the probability mass functions of the photon events
        model_name (str): describes the model used for fitting
    Returns:
        (array): rectangular array with the difference between the calculated model and pd_data
    """
    (dim,) = pd_data.shape
    n_modes = int(params["n_modes"])
    sq_n = [params["sq_n" + str(i)] for i in range(n_modes)]
    eta = params["eta"]
    n_dark = params["n_dark"]
    model_pmf = degenerate_pmf(n_max, eta=eta, sq_n=sq_n, n_dark=n_dark)[0:dim]

    return model_pmf - pd_data


def fit_1d(model_name, pd_data, guess, method="leastsq"):
    """Takes as input the name of the model to fit to and the jpd of the data
    and returns the fitted model.
    Args:
        model_name (str): describes the model used for fitting
        pd_data (array): one dimensional array of the probability distribution of the data
        guess (dict): dictionary with the guesses for the different parameters
    Returns:
        Object containing the optimized parameter and several goodness-of-fit statistics
    """
    pars_model = Parameters()
    n_modes = guess["n_modes"]
    pars_model.add("n_modes", value=n_modes, vary=False)
    # Add the squeezing parameters
    for i in range(n_modes):
        pars_model.add("sq_n" + str(i), value=guess["sq_n" + str(i)], min=0.0)

    if model_name == "NoiseFixed":
        pars_model.add("eta", value=guess["eta"], min=0.0, max=1.0)
        pars_model.add("n_dark", value=guess["n_dark"], vary=False)
    elif model_name == "LossFixed":
        pars_model.add("eta", value=guess["eta"], vary=False)
        pars_model.add("n_dark", value=guess["n_dark"], min=0.0)
    elif model_name == "NoiseLossFixed":
        pars_model.add("eta", value=guess["eta"], vary=False)
        pars_model.add("n_dark", value=guess["n_dark"], vary=False)
    elif model_name == "NoneFixed":
        pars_model.add("eta", value=guess["eta"], min=0.0, max=1.0)
        pars_model.add("n_dark", value=guess["n_dark"], min=0.0)
    else:
        raise NotImplementedError("Model " + model_name + " not implemented.")

    minner_model = Minimizer(model_1d, pars_model, fcn_args=([pd_data]))
    result_model = minner_model.minimize(method=method)

    return result_model


sqtom/fitting_2d.py
# Copyright 2019-2020 Xanadu Quantum Technologies Inc.

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#     http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
Twin-beam squeezer inverse-problem solver
=========================================
This module solves the *inverse* problem: given a joint photon number distribution find the best
parameters describing the different quantum states in a twin-beam producing it.

The ideas behind this module borrow heavily on the work of Burenkok et al. in

Full statistical mode reconstruction of a light field via a photon-number-resolved measurement
I. A. Burenkov, A. K. Sharma, T. Gerrits, G. Harder, T. J. Bartley, C. Silberhorn, E. A. Goldschmidt, and S. V. Polyakov
Phys. Rev. A 95, 053806 (2017)
"""

import numpy as np
from lmfit import Minimizer, Parameters
from sqtom.forward_solver import twinbeam_pmf


def two_schmidt_mode_guess(jpd_data, sq_label="sq_", noise_label="noise"):
    """Given a two mode histogram, this function generates a "physically" motivated guess for the loss, Schmidt occupations
    and dark counts parameters.

    This model is sensible only if the average g2 of signal and idler is above 1.5.

    Args:
        jpd_data (array): rectangular array with the probability mass functions of the photon events
        sq_label (string): label for the squeezing parameters.
        noise_label (string): label for the noise parameters.

    Returns:
        dict: dictionary containing a set of "reasonable" model parameters
    """
    res = marginal_calcs_2d(jpd_data)
    g2avg = np.max([0.5 * (res["g2_s"] + res["g2_i"]), 1.5])
    Nbar = 1.0 / (res["g11"] - g2avg)
    n0 = 0.5 * (1 + np.sqrt(2 * g2avg - 3)) * Nbar
    n1 = 0.5 * (1 - np.sqrt(2 * g2avg - 3)) * Nbar
    etas = res["n_s"] / Nbar
    etai = res["n_i"] / Nbar
    noise = np.abs(res["g2_s"] - res["g2_i"]) * Nbar
    return {
        "eta_s": etas,
        "eta_i": etai,
        sq_label + "0": n0,
        sq_label + "1": n1,
        noise_label + "_s": noise * etas,
        noise_label + "_i": noise * etai,
        "n_modes": 2,
    }


def marginal_calcs_2d(jpd_data, as_dict=True):
    """Given a two dimensional array of probabilities it calculates
    the mean photon numbers, their g2's and g11.

    It returns these values as a dictionary or as an array.

    Args:
        jpd_data (array): probability mass function of the photon events
        as_dict (boolean): whether to return the results as a dictionary

    Returns:
        dict or array: values of the mean photons number for signal and idlers,
            their corresponding g2, their g11 and their noise reduction factor (nrf).
    """
    inta, intb = jpd_data.shape
    na = np.arange(inta)
    nb = np.arange(intb)
    ps = np.sum(jpd_data, axis=1)
    pi = np.sum(jpd_data, axis=0)
    ns = ps @ na
    ni = pi @ nb
    ns2 = ps @ (na ** 2)
    ni2 = pi @ (nb ** 2)
    ns3 = ps @ (na ** 3)
    ni3 = pi @ (nb ** 3)
    g2s = (ns2 - ns) / ns ** 2
    g2i = (ni2 - ni) / ni ** 2
    g3s = (ns3 - 3 * ns2 + 2 * ns) / ns ** 3
    g3i = (ni3 - 3 * ni2 + 2 * ni) / ni ** 3
    g11 = (na @ jpd_data @ nb) / (ns * ni)
    nrf = np.sum([[((i - j) ** 2) * jpd_data[i, j] for i in range(inta)] for j in range(intb)])
    nrf -= np.sum([[(i - j) * jpd_data[i, j] for i in range(inta)] for j in range(intb)]) ** 2
    nrf /= ns + ni
    if as_dict is True:
        return {
            "n_s": ns,
            "n_i": ni,
            "g11": g11,
            "g2_s": g2s,
            "g2_i": g2i,
            "nrf": nrf,
            "g3_s": g3s,
            "g3_i": g3i,
        }
    return np.array([ns, ni, g11, g2s, g2i, nrf, g3s, g3i])


def gen_hist_2d(beam1, beam2):
    """Calculate the joint probability mass function of events.

    Args:
        beam1 (array): 1D events array containing the raw click events of first beam
        beam2 (array): 1D events array containing the raw click events of second beam

    Returns:
        array: probability mass function of the click patterns for the two beams
    """
    nx = np.max(beam1)
    ny = np.max(beam2)
    xedges = np.arange(nx + 2)
    yedges = np.arange(ny + 2)
    mass_fun, _, _ = np.histogram2d(beam1, beam2, bins=(xedges, yedges), normed=True)
    return mass_fun


def threshold_2d(ps, nmax, mmax):
    """Thresholds a 2D probability distribution by assigning events with more than nmax (mmax) photons
    in the first (second) axis to the nmax (mmax) bin.

    Args:
        ps (array): probability distribution
        nmax (int): threshold for first axis
        mmax (int): threshold for second axis

    Returns:
        array: thresholded probability distribution
    """
    probs = np.copy(ps[:nmax, :mmax])
    for i in range(mmax - 1):
        probs[nmax - 1, i] += np.sum(ps[nmax:, i])
    for i in range(nmax - 1):
        probs[i, mmax - 1] += np.sum(ps[i, mmax:])
    probs[nmax - 1, mmax - 1] = np.sum(ps[nmax - 1 :, mmax - 1 :])
    return probs


def fit_2d(
    pd_data,
    guess,
    do_not_vary=None,
    method="leastsq",
    threshold=False,
    cutoff=50,
    sq_label="sq_",
    noise_label="noise",
):
    """Returns a model fit from the parameter guess and the data

    Args:
        pd_data (array): one dimensional array of the probability distribution of the data
        guess (dict): dictionary with the guesses for the different parameters
        method (string): method to be used by the optimizer
        threshold (boolean or list): threshold photon number for the signal and idlers
        do_not_vary (list): list of variables that should be held constant during optimization
        cutoff (int): internal cutoff
        sq_label (string): label for the squeezing parameters.
        noise_label (string): label for the noise parameters.

    Returns:
        Object: object containing the optimized parameter and several goodness-of-fit statistics
    """
    if do_not_vary is None:
        do_not_vary = []
    pars_model = Parameters()
    n_modes = guess["n_modes"]
    pars_model.add("n_modes", value=n_modes, vary=False)
    for i in range(n_modes):
        pars_model.add(sq_label + str(i), value=guess["sq_" + str(i)], min=0.0)

    if "eta_s" in do_not_vary:
        pars_model.add("eta_s", value=guess["eta_s"], vary=False)
    else:
        pars_model.add("eta_s", value=guess["eta_s"], min=0.0, max=1.0)

    if "eta_i" in do_not_vary:
        pars_model.add("eta_i", value=guess["eta_i"], vary=False)
    else:
        pars_model.add("eta_i", value=guess["eta_i"], min=0.0, max=1.0)

    if noise_label + "_s" in do_not_vary:
        pars_model.add(noise_label + "_s", value=guess[noise_label + "_s"], vary=False)
    else:
        pars_model.add(noise_label + "_s", value=guess[noise_label + "_s"], min=0.0)

    if noise_label + "_i" in do_not_vary:
        pars_model.add(noise_label + "_i", value=guess[noise_label + "_i"], vary=False)
    else:
        pars_model.add(noise_label + "_i", value=guess[noise_label + "_i"], min=0.0)

    if threshold:

        def model_2d(params, jpd_data):
            (dim_s, dim_i) = pd_data.shape
            joint_pmf = twinbeam_pmf(params, cutoff=cutoff)
            return threshold_2d(joint_pmf, dim_s, dim_i) - threshold_2d(jpd_data, dim_s, dim_i)

    else:

        def model_2d(params, jpd_data):
            (dim_s, dim_i) = pd_data.shape
            joint_pmf = twinbeam_pmf(params, cutoff=cutoff)[:dim_s, :dim_i]
            return joint_pmf - jpd_data

    minner_model = Minimizer(model_2d, pars_model, fcn_args=([pd_data]))
    result_model = minner_model.minimize(method=method)

    return result_model
